### ISSUE TITLE
rpm install failed：error: unpacking of archive failed on file /usr/sbin/aaa;636c6826: cpio: open failed - No such file or directory

### ISSUE THREAD

**Issue by JetXujing** [rpm-software-management/rpm#2273]
1. rpmbuild --with=sbinsymlinks -bb test.spec
```
[root@localhost SPECS]# cat test.spec
%bcond_without sbinsymlinks

Name: test
Version: 1.0
Release: 1
License: MIT
Summary: Test
BuildArch: noarch

%description
Test

%prep

%build

%install
mkdir -p %{buildroot}/usr/sbin/
touch %{buildroot}/usr/sbin/aaa

%if %{with sbinsymlinks}
mkdir %{buildroot}/sbin/
cd %{buildroot}/sbin/
ln -sv ..%{_sbindir}/aaa .
%endif

%post

%files
/usr/sbin/aaa

%if %{with sbinsymlinks}
/sbin/aaa
%endif
```
2. rpm -i /root/rpmbuild/RPMS/noarch/test-1.0-1.noarch.rpm --force

when `sbin -> usr/sbin` soft link exists, according to the preceding steps, the following error information is displayed:
```
[root@localhost SPECS]# rpm -i /root/rpmbuild/RPMS/noarch/test-1.0-1.noarch.rpm --force
error: unpacking of archive failed on file /usr/sbin/aaa;636c693f: cpio: open failed - No such file or directory
error: test-1.0-1.noarch: install failed
```

The problem is caused by the a82251b44ee2d2802ee8aea1b3d89f88beee4bad submission. The cause is that a temporary file /usr/sbin/aaa;xxx is created using openat(), and then the file /sbin/aaa;xxx is created using openat(), openat() failed because the file already exists. Before the a82251b44ee2d2802ee8aea1b3d89f88beee4bad is integrated, the temporary file is renamed and then the next temporary file is generated. Therefore, there's no open same files case.

**Comment by JetXujing** [comment]
When I tried to fix it with the following patch, the `openat(dirfd, dest, O_WRONLY|O_CREAT, 0200)` still failed with a message "No such file or directory". In this case, the open does not seem to cover this scenario.
```
diff --git a/lib/fsm.c b/lib/fsm.c
index c9ab3e1..4ceca3d 100644
--- a/lib/fsm.c
+++ b/lib/fsm.c
@@ -282,6 +282,12 @@ static int fsmOpen(int *wfdp, int dirfd, const char *dest)
     /* Create the file with 0200 permissions (write by owner). */
     int fd = openat(dirfd, dest, O_WRONLY|O_EXCL|O_CREAT, 0200);
 
+    if (fd < 0 && errno == EEXIST) {
+	rpmlog(RPMLOG_WARNING, "xujing: %8s (%s [%d]) %s\n", __func__,
+	       dest, fd, (rc < 0 ? strerror(errno) : ""));
+	fd = openat(dirfd, dest, O_WRONLY|O_CREAT, 0200);
+    }
+
     if (fd < 0)
 	rc = RPMERR_OPEN_FAILED;
```

**Comment by JetXujing** [comment]
I tried to modify the %install and the installation still failed, but it was due to symlink failure.
```
%install
mkdir %{buildroot}/sbin/
touch %{buildroot}/sbin/aaa

%if %{with sbinsymlinks}
mkdir -p %{buildroot}/usr/sbin/
cd %{buildroot}/usr/sbin/
ln -sv ../../sbin/aaa .
%endif
```
```
[root@localhost SPECS]# rpm -ivh /root/rpmbuild/RPMS/noarch/test-1.0-1.noarch.rpm --force
Verifying...                          ################################# [100%]
Preparing...                          ################################# [100%]
Updating / installing...
   1:test-1.0-1                       ################################# [100%]
error: unpacking of archive failed on file /usr/sbin/aaa;639a7f70: cpio: **symlink failed - No such file or directory**
error: test-1.0-1.noarch: install failed
```

**Comment by pmatilai** [comment]
I can't reproduce the exact error but at least on Fedora where /sbin is symlinked to /usr/sbin/ on directory level, this is what I get:

> [root@localhost ~]# rpm -ivh /tmp/test-1.0-1.noarch.rpm 
Verifying...                          ################################# [100%]
Preparing...                          ################################# [100%]
	file /usr/sbin/aaa conflicts between attempted installs of test-1.0-1.noarch and test-1.0-1.noarch
[root@localhost ~]# 

In other words, the package conflicts with itself due to the directory level symlink. And if you now try to force the issue, it will fail and should quite rightly do so. Never use --force, really.

**Comment by pmatilai** [comment]
So, given the error is not exactly the same in my case - what is the behavior you get *without* --force?

**Comment by JetXujing** [comment]
I get the same results as you, without -- force.
```
[root@localhost ~]# rpm -ivh /tmp/test-1.0-1.noarch.rpm
Verifying... ################################# [100%]
Preparing... ################################# [100%]
file /usr/sbin/aaa conflicts between attempted installs of test-1.0-1.noarch and test-1.0-1.noarch
[root@localhost ~]#
```
The problem that I described arises when --force is used.

**Comment by pmatilai** [comment]
Ack, thanks for confirming that.

The "use-case" for --force is to replace some file(s) from another package with different content, in which case it's reasonable to allow an operator override. But there's no circumstance in which it makes sense for a package to overwrite it's own contents as would happen here. So ideally rpm should differentiate between conflict types and only let --force influence things that can be meaningfully achieved...
